core/jvm/src/main/scala/fs2/interop/flow/StreamSubscription.scala

```diff
@@ -39,35 +40,53 @@ import java.util.concurrent.Flow.{Subscription, Subscriber}
 private[flow] final class StreamSubscription[F[_], A] private (
     stream: Stream[F, A],
     sub: Subscriber[A],
-    requestDispatcher: Dispatcher[F],
-    requests: Queue[F, StreamSubscription.Request],
-    canceled: Deferred[F, Unit]
+    requests: AtomicLong,
+    resume: AtomicReference[() => Unit],
+    canceled: AtomicReference[() => Unit]
 )(implicit F: Async[F])
     extends Subscription {
-  // Ensure we are on a terminal state; i.e. set `canceled`, before signaling the subscriber.
-  private def onError(ex: Throwable): F[Unit] =
-    cancelMe >> F.delay(sub.onError(ex))
+  // Ensure we are on a terminal state; i.e. call `cancel`, before signaling the subscriber.
+  private def onError(ex: Throwable): Unit = {
+    cancel()
+    sub.onError(ex)
+  }
 
-  private def onComplete: F[Unit] =
-    cancelMe >> F.delay(sub.onComplete())
+  private def onComplete(): Unit = {
+    cancel()
+    sub.onComplete()
+  }
 
   private[flow] def run: F[Unit] = {
-    def subscriptionPipe: Pipe[F, A, A] =
-      in => {
-        def go(s: Stream[F, A]): Pull[F, A, Unit] =
-          Pull.eval(requests.take).flatMap {
-            case StreamSubscription.Request.Infinite =>
-              s.pull.echo
-
-            case StreamSubscription.Request.Finite(n) =>
+    val subscriptionPipe: Pipe[F, A, A] = in => {
+      def go(s: Stream[F, A]): Pull[F, A, Unit] =
+        Pull.eval(F.delay(requests.get())).flatMap { n =>
+          if (n == Long.MaxValue)
+            s.pull.echo
+          else if (n == 0)
+            Pull.eval(F.asyncCheckAttempt[Unit] { cb =>
+              F.delay {
+                // If there aren't more pending request,
+                // we will wait until the next one.
+                resume.set(() => cb.apply(Either.unit))
+                // However, before blocking,
+                // we must check if it has been a concurrent request.
+                // In case it was, we abort the wait.
+                if (requests.get() > 0)
+                  Either.unit
```

Comment:
If we abort, should we clear `resume` of the callback? At the very least to avoid a memory leak.